articles/150_roslaunch.md

```diff
@@ -0,0 +1,364 @@
+
+#### Execution
+
+Managed ROS Nodes do not add any additional inputs or specific configurations at execution time on top of what "plain" ROS nodes add, at least not at this time.
+In the future this might change, so reference the design doc[^lifecylce] or future documentation on the subject.
+
+#### Runtime
+
+During runtime, a Managed ROS node emits events anytime the state of the node changes.
+This is at least emitted on a topic, but could also be captured, aggregated, and/or communicated in other ways too.
+These state changes could be consumed by either the launch system itself or by the user, either of which could react to these changes.
+
+For example, the user could express something like "when node 'A' enters the `Active` state, launch nodes 'B' and 'C'" or "if node 'A' exits with a return code or enters the `Finalized` state, shutdown everything".
+
+<div class="alert alert-warning" markdown="1">
+RFC:
+
+This is where things get murky for me.
+There is a gray line between more complex startup behaviors (like "wait for A to get to Active then launch B") and long running monitoring and reactions to events, what I would call "supervision" (like "if A goes transitions to ErrorProcessing change B to Inactive").
+
+I'm interested to see what other people think.
+
+For me the best way to separate the two cases (if they're to be separated at all) is to ask "do I want/need to understand this event handling off-line"?
+For example, when looking at a system description, it would be nice to see the event handler "when A crashes or exits, the whole system is down" (i.e. `require=true`) or the event handler "when A reaches Inactive, launch B" (something like "B depends on A").
+However, maybe it doesn't need to capture/understand the "when A crashes or has an error move B to inactive".
+
+Alternatively, we could say "all event handling is opaque to the launch system and is user defined" which removes the ambiguity.
+
+\<wjwwod's opinion>
+<br/>
+I think the idea that saying all "supervision" or "reactionary launching" like activities are not part of the launch system is the simplest way forward, but then you'd still like to include in the launch system some convenient ways to express the most common cases, e.g. something like ROS 1's `roslaunch`'s `require=true` and `respawn=true`.
+In my opinion, it would be unacceptably inconvenient to have users write a lambda for each of those cases, and I also think it would lead to everyone doing it slightly different and making it impossible to understand with tools.
+<br/>
+</wjwwod's opinion>
+
+This will come up again in the "System Description" section, because I believe the mind set around the "Execution of the System Description" (basically the python library that runs things) will be "personal responsibility".
```

Comment:
Is there a general ROS 2 design decision to move away from the ROS 1 concept that each node should be written in a way that is independent of the order in which other nodes start?

Many people question that approach, but experience suggests that it was a good way to write reusable nodes.

If not, then the use case of "wait for A to get to Active then launch B" may not be very important. It plays into a simple-minded system design paradigm that we may not want to encourage.

Reply:
To me, the more important problem is providing recovery, availability and serviceability in the face of node or process failures. That could involve:
* if a node fails, restart it
  - what do other nodes do in the interim? Just wait?
* if a node cannot be restarted (device failure perhaps?), run something else as a fallback
* if everything is FUBAR, capture as much relevant information as possible for off-line analysis

Reply:
+1 for focusing on recovery, availability and serviceability in the face of surviving post-apocolyptic failures.

One item that would have been very useful is for roslaunch to report out (with relevant stdout/stderr) on launch failures or crashes, either to a watchdog node or a graphical interface to the robot. We used to track a node's existence from watchdog programs, but were not able to extract stdout/stderr.

Perhaps for this, a useful feature would be the ability to provide callbacks for launch failures/crashes? This could enable a variety of patterns, such as some of those suggested by Jack above. Common use cases could be implemented as extensions later.

Reply:
I don't think the intention is to discourage writing nodes which are decoupled and asynchronous, just that in a few situations having a coordinated startup is very important and in ROS 1 the system worked against you trying to do that. Hopefully the new launch system provides the ability to have some more sophisticated startup behavior without changing the common practice or having nodes which are decoupled from their peers.

It is quite a bit harder to write launch files (in Python or the new XML frontend) which operate in an "if this then that" manner, so I don't think people will default to that pattern, instead choosing only to use it if they need to do so.